### ISSUE TITLE
Expectations pass if called in the wrong order

### ISSUE THREAD

**Issue by SystemParadox** [cjohansen/Sinon.JS#374]
I am trying to test a subscription system which takes an object of callbacks, but expectations don't seem to preserve any kind of ordering.

```
var foo = {
    bar: function (val) { },
    baz: function (val) { }
};
var mock = sinon.mock(foo);
mock.expects('bar').withArgs(1);
mock.expects('bar').withArgs(2);
foo.bar(2);
foo.bar(1);
mock.verify();
```

Expected result: Exception, test failure.

Actual result: Nothing, the test passes without any errors.

I notice in other discussions that this used to work, but was deliberately changed? I don't understand why. That being the case, what is the recommended way to test a specific sequence of actions?

The only way I can see to accomplish this feels a bit clumsy:

```
var foo = {
    bar: sinon.spy(),
    baz: sinon.spy(),
};
foo.bar(2);
foo.bar(1);
assert.callCount(foo.bar, 2);
assert.callCount(foo.baz, 0);
assert(foo.bar.getCall(0).calledWith(1));
assert(foo.bar.getCall(1).calledWith(2));
```

I also have the need to test cases where calls to bar and baz should be interleaved, but I'm not sure that this is even possible using separate spies.

Any thoughts?

Thanks.

**Comment by mantoni** [comment]
We are trying to keep the GitHub issues list tidy and focused on bugs and feature discussions. This ticket looks like a usage question, please post it to the [Sinon.JS mailinglist](http://groups.google.com/group/sinonjs), so the bigger community can help answer your questions.

If you feel that your topic is an issue with Sinon.JS, please open a new ticket and follow the [guidelines for reporting an issue](https://github.com/cjohansen/Sinon.JS/blob/master/CONTRIBUTING.md#reporting-an-issue).

**Comment by SystemParadox** [comment]
This is a bug/feature discussion. Expectations should enforce call-order.

**Comment by mantoni** [comment]
Sorry, I should have explained in more detail. This issue has been inactive for too long.

Changing this behavior to enforce call order verification is a breaking change. It's also not obvious to users if we would put this in. Therefore we are not going to change it.

I posted a couple of examples on how you can verify call orders with Sinon. I hope they help.

If you have ideas on how this could be added in without changing the existing behavior, please open a new issue.

Thanks.

**Comment by pronebird** [comment]
@mantoni mailing list 😆 

I think the issue deserves more attention and it doesn't matter if it's a breaking change or not since it's a relevant use case.

Expectations may implement some `.then()` to automatically do all the calls to `.callOrder` behind the scenes.

**Comment by pronebird** [comment]
@mantoni the last time I checked it was sinon 3.0 so things probably changed since then, that's why I ask. 

However I find documentation hard to follow because there are no types so I can't know what for instance `mock.expects('method').once().withArgs(..)` returns. So again I don't understand how to ensure the call order.

**Comment by mantoni** [comment]
Just to set expectations right: I don't think mocks are going to receive much love from the Sinon team in the future. There have been talks about moving them into a separate project and out of the Sinon core. We're simply not using them anymore. You may still send a PR if you want to look into it yourself.

In case you want to work with the `fake`, `spy` or `stub` APIs, you have these possibilities:

- `sinon.assert.callOrder(spy1, spy2, ...)` - see http://sinonjs.org/releases/v6.0.0/assertions/
- `spyCall.calledBefore(otherCall)`, `spyCall.calledAfter(otherCall)`, `spyCall.calledImmediatelyBefore(otherCall)` and `spyCall.calledImmediatelyAfter(otherCall)` - see http://sinonjs.org/releases/v6.0.0/spy-call/

From my understanding, these APIs should also work with mock calls, but not entirely sure.

**Comment by pronebird** [comment]
Alright, thanks. I use a very tiny helper to achieve this but without `sinon` and then we have a lot of `chai` in tests historically. I was thinking to bring in more of `sinon` to replace our helper, but I guess given the circumstance that's not going to happen.
